### PR TITLE
Astro alerts use case

### PR DESCRIPTION
**Pull request by kentdanas**
New astro use case doc for alerting, meant to address customer questions like:
- When should I use Astro alerts vs built-in Airflow alerts?
- How can I implement a combination of Airflow and Astro alerts?
- Why are Astro SLAs better than Airflow SLAs?

### CODE REVIEW

File: astro/use-cases/airflow-astro-alerting.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+This use case depends on the following Astro and Airflow features to create different types of alerts for your pipelines:
+
+- Configuring DAG SLAs and failure notifications using [Astro alerts](alerts.md).
+- Implementing [Airflow callbacks](https://docs.astronomer.io/learn/error-notifications-in-airflow#airflow-callbacks) for custom task-level alerts. 
+
+## Prerequisites
+
+This use case assumes you have:
+
+- At least one [Astro Deployment](create-deployment). Your deployment must run Astro Runtime 7.1.0 or later and it must have OpenLineage enabled.
```
Comment (TJaniF) [inline astro/use-cases/airflow-astro-alerting.md:30]:
```suggestion
- At least one [Astro Deployment](create-deployment.md). Your deployment must run Astro Runtime 7.1.0 or later and it must have OpenLineage enabled.
```

File: astro/use-cases/airflow-astro-alerting.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+## Explanation
+
+Using a combination of Astro alerts and Airflow notifications allows your team to easily implement alerting logic for any scenario. You may need different types of alerts that are better suited towards one option or the other, as shown in the example in this use case:
+
+- Task-level alerts with custom logic, for example if you want to be notified of success or failure only for a specific task, or want to run custom code if a task succeeds or fails, are straightforward to implement in Airflow. Task-level success or failure alerts or alerts with custom logic are not currently supported with Astro alerts.
+- DAG-level success or failure alerts can be accomplished using either Airflow notifications or Astro alerts, but on Astro they are easy to configure and do not require any code changes. To implement these alerts with Airflow, you would need to update your DAG code and possibly complete more configuration for your communication channel (e.g. setting up an SMTP server for email alerts).
+- Absolute time alerts are straightforward to implement with Astro alerts, but are not currently supported in Airflow.
```
Comment (virajmparekh) [inline astro/use-cases/airflow-astro-alerting.md:51]: Maybe reword this as: "Absolute time alerts, or alerts where you don't want to wait for the task's state to change, should be done via the Astro alerts." Stylistic choice

**Review by virajmparekh — COMMENTED** [review]
Few small things, but otherwise another great doc!

File: astro/use-cases/airflow-astro-alerting.md (hunk 2 above)
Comment (kentdanas) [reply astro/use-cases/airflow-astro-alerting.md:51]: What do you mean by "wait for the task's state to change"? Isn't that the case with any SLA? Regardless, how about "Absolute time alerts should be implemented using Astro alerts. They are not currently supported in Airflow."

Want to be clear that Airflow doesn't have this mechanism today.

File: astro/use-cases/airflow-astro-alerting.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+- The Astro alert for absolute time allows you to implement alerts based on a specific time of day your DAG should complete. This type of SLA is not available in Airflow.
+- Astro alerts allow you to easily choose between multiple common communication channels. With Airflow SLAs, if you want your notification sent somewhere other than your email, you have to write your own custom callback logic. Additionally, if you have an SMTP service configured in your Airflow environment you will get email notifications for SLA misses; there is no way to turn this behavior off.
+
+### Recommended alerts
+
+Generally, Astronomer recommends choosing the following alerts for these common scenarios:
+
+| Scenario                                   | Astro alert | Airflow email notification | Airflow callback | Airflow SLA | Airflow timeout |
+|--------------------------------------------|-------------|-----------------------------|------------------|--------------|------------------|
+| Email on DAG success or retry               | X           | X                           |                  |              |                  |
```
[the remark this answers is not in the document]
Comment (kentdanas) [reply astro/use-cases/airflow-astro-alerting.md:73]: Because until we have a programmatic way of creating alerts, this one could truly go either way. I have a paragraph somewhere explaining this, but if you were moving on to Astro and had 100 DAGs with email alerts already set up, it's definitely easier to just set up SMTP on Astro than to create 100 different Astro alerts. If you have a smaller number of DAGs, or don't already have alerts configured, then Astro alerts are preferable. Once adding bulk alerts via API or something like that is available in the product, I'd be more comfortable suggesting Astro alerts explicitly for this case.

File: astro/use-cases/airflow-astro-alerting.md (hunk 2 above)
Comment (kentdanas) [inline astro/use-cases/airflow-astro-alerting.md:51]:
```suggestion
- Absolute time alerts, for when you need your DAG to complete by a certain time of day, should be implemented using Astro alerts. They are not currently supported in Airflow
```

**Comment by kentdanas** [comment]
@jwitz @lzdanski I'm waiting on a SME review for this, but in the meantime I don't expect huge changes so I think we're okay to go ahead with editing as well.

File: astro/use-cases/airflow-astro-alerting.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+- At least one [Astro Deployment](https://docs.astronomer.io/astro/create-deployment). Your deployment must run Astro Runtime 7.1.0 or later and it must have OpenLineage enabled.
+- An [Astro project](https://docs.astronomer.io/astro/cli/develop-project) with at least one DAG. Your DAG should have at least two tasks.
+
+However, you can extend this use case to encompass any number of Astro Deployments and DAGs.
+
+## Implementation
+
+To implement this use case:
+
+1. In the DAG in your Astro project, add an `on_success_callback` that uses the pre-built SlackNotifier to one task (not the DAG's `default_args`). See [Example pre-built notifier: Slack](https://docs.astronomer.io/learn/error-notifications-in-airflow#example-pre-built-notifier-slack).
```
Comment (jwitz) [inline astro/use-cases/airflow-astro-alerting.md:39]:
```suggestion
1. Open the DAG in your Astro project and configure a pre-built SlackNotifier for one of its tasks as a task-level argument. See [Example pre-built notifier: Slack](https://docs.astronomer.io/learn/error-notifications-in-airflow#example-pre-built-notifier-slack) for sample code.  
```

File: astro/use-cases/airflow-astro-alerting.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+- An [Astro project](https://docs.astronomer.io/astro/cli/develop-project) with at least one DAG. Your DAG should have at least two tasks.
+
+However, you can extend this use case to encompass any number of Astro Deployments and DAGs.
+
+## Implementation
+
+To implement this use case:
+
+1. In the DAG in your Astro project, add an `on_success_callback` that uses the pre-built SlackNotifier to one task (not the DAG's `default_args`). See [Example pre-built notifier: Slack](https://docs.astronomer.io/learn/error-notifications-in-airflow#example-pre-built-notifier-slack).  
+2. Deploy your project to your Astro Deployment. See [Deploy code to Astro](https://docs.astronomer.io/astro/deploy-code).
```
Comment (jwitz) [inline astro/use-cases/airflow-astro-alerting.md:40]:
```suggestion
2. Deploy your project to your Astro Deployment. See [Deploy code to Astro](deploy-code.md).
```

File: astro/use-cases/airflow-astro-alerting.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+3. Add a connection to Slack in the Astro Cloud UI. See [Create Airflow connections in the Cloud UI](https://docs.astronomer.io/astro/create-and-link-connections). This connection will be used by your SlackNotifier. Make sure the connection ID matches what you used in your DAG code in Step 1.
```
Comment (jwitz) [inline astro/use-cases/airflow-astro-alerting.md:41]:
```suggestion
3. Add a connection to Slack in the Astro Cloud UI. See [Create Airflow connections in the Cloud UI](create-and-link-connections.md). This connection will be used by your SlackNotifier, so make sure the connection ID matches what you used in your DAG code in Step 1.
```

File: astro/use-cases/airflow-astro-alerting.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+4. Configure an Astro DAG failure alert for your DAG using the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).
```
Comment (jwitz) [inline astro/use-cases/airflow-astro-alerting.md:42]:
```suggestion
4. In your Astro Deployment, configure an Astro **DAG failure** alert for your DAG using the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).
```

File: astro/use-cases/airflow-astro-alerting.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+5. Configure an Astro absolute time alert for your DAG based on the amount of time you expect your DAG to complete in. Use the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).
```
Comment (jwitz) [inline astro/use-cases/airflow-astro-alerting.md:43]:
```suggestion
5. Configure an Astro **Absolute time** alert for your DAG based on the amount of time you expect your DAG to complete in. Use the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).
```

File: astro/use-cases/airflow-astro-alerting.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+## Explanation
+
+Using a combination of Astro alerts and Airflow notifications allows your team to easily implement alerting logic for any scenario. You may need different types of alerts that are better suited towards one option or the other, as shown in the example in this use case:
```
Comment (jwitz) [inline astro/use-cases/airflow-astro-alerting.md:47]:
```suggestion
Using a combination of Astro alerts and Airflow notifications allows your team to implement alerting logic using the best tool for each scenario. You might need different types of alerts that are better suited towards one option or the other, as shown in the example in this use case:
```

File: astro/use-cases/airflow-astro-alerting.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+- Task-level alerts with custom logic, for example if you want to be notified of success or failure only for a specific task, or want to run custom code if a task succeeds or fails, are straightforward to implement in Airflow. Task-level success or failure alerts or alerts with custom logic are not currently supported with Astro alerts.
+- DAG-level success or failure alerts can be accomplished using either Airflow notifications or Astro alerts, but on Astro they are easy to configure and do not require any code changes. To implement these alerts with Airflow, you would need to update your DAG code and possibly complete more configuration for your communication channel (e.g. setting up an SMTP server for email alerts).
+- Absolute time alerts, for when you need your DAG to complete by a certain time of day, should be implemented using Astro alerts. They are not currently supported in Airflow
```
Comment (jwitz) [inline astro/use-cases/airflow-astro-alerting.md:51]:
```suggestion
- **Task-level alerts with custom logic** are recommended if you want be notified of success or failure only for a specific task, or you want to run custom code if a task succeeds or fails. Task-level success or failure alerts or alerts with custom logic are not currently supported with Astro alerts, but they're straightforward to implement in Airflow
- DAG-level success or failure alerts can be created using either Airflow notifications or Astro alerts, but on Astro they are easy to configure and don't require any code changes. To implement these alerts with Airflow, you need to update your DAG code and complete additional configuration for your communication channel (for example, setting up an SMTP server for email alerts).
- Absolute time alerts are recommended when you need your DAG to complete by a certain time of day. These types of alerts are available only on Astro.
```

File: astro/use-cases/airflow-astro-alerting.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+In some cases, whether you set up an alert on Astro or in Airflow will be dependent on how your DAGs were configured before you moved to Astro. For example, if all of your DAGs have email on failure notifications configured already, it will likely be easier for you to configure an SMTP server on Astro and rely on the existing notification rather than implementing a new Astro alert for each DAG. On the other hand, if you are developing new DAGs on Astro, it will likely be easier to implement Astro alerts which require less configuration for the communication channels.
```
Comment (jwitz) [inline astro/use-cases/airflow-astro-alerting.md:53]:
```suggestion
In some cases, whether you set up an alert on Astro or in Airflow will depend on how your DAGs were configured before you moved to Astro. For example, if all of your DAGs already have email-on-failure notifications configured, it will likely be easier for you to configure an SMTP server on Astro and rely on the existing notification than implement a new Astro alert for each DAG. On the other hand, if you are developing new DAGs on Astro, it will likely be easier to implement Astro alerts which require less configuration for the communication channels.
```

File: astro/use-cases/airflow-astro-alerting.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+:::note
```
Comment (jwitz) [inline astro/use-cases/airflow-astro-alerting.md:55]:
```suggestion
:::info
```

File: astro/use-cases/airflow-astro-alerting.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+Programmatically creating Astro alerts, which may be helpful if you have a large number of DAGs, is coming soon.
+
+:::
+
+In other cases, Astro alerts are clearly the better option. If you need to implement SLAs for your pipelines so you know as soon as a task is taking too long or a DAG hasn't completed on time, you should use the absolute time or task duration Astro alerts. There are a couple of reasons these are preferable to Airflow SLAs:
```
Comment (jwitz) [inline astro/use-cases/airflow-astro-alerting.md:61]:
```suggestion
In other cases, Astro alerts are always the better option. If you need to implement SLAs for your pipelines so you immediately know when a DAG task is taking too long to run, you should use the absolute time or task duration Astro alerts. There are a couple of reasons these are preferable to Airflow SLAs:
```

**Review by jwitz — APPROVED** [review]
Awesome use case! Breaks down all available options in a really accessible and actionable way. I just had some minor wording suggestions throughout, please tag me back if the title's not quite right - I think we can make it more specific to the use case so that it's distinguishable from the core Astro alerts doc

File: astro/use-cases/airflow-astro-alerting.md (hunk 5 above)
Comment (kentdanas) [reply astro/use-cases/airflow-astro-alerting.md:40]: @jwitz I had this before but the build failed because it couldn't resolve the link - I'm assuming because it's in a different folder? Not sure if there's a way around that?

File: astro/use-cases/multi-channel-alerting-systems.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+4. In your Astro Deployment, configure an Astro **DAG failure** alert for your DAG using the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).
+5. Configure an Astro **Absolute time** alert for your DAG based on the amount of time you expect your DAG to complete in. Use the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).
+
+## Explanation
+
+Using a combination of Astro alerts and Airflow notifications allows your team to implement alerting logic using the best tool for each scenario. You might need different types of alerts that are better suited towards one option or the other, as shown in the example in this use case:
+
+- **Task-level alerts with custom logic** are recommended if you want be notified of success or failure only for a specific task, or you want to run custom code if a task succeeds or fails. Task-level success or failure alerts or alerts with custom logic are not currently supported with Astro alerts, but they're straightforward to implement in Airflow
```
Comment (paolaperaza) [inline astro/use-cases/multi-channel-alerting-systems.md:49]:
```suggestion
- **Task-level alerts with custom logic** are recommended if you want be notified of success or failure only for a specific task, or you want to run custom code if a task succeeds or fails. Task-level success or failure alerts or alerts with custom logic are not currently supported with Astro alerts, but they're straightforward to implement in Airflow.
```

File: astro/use-cases/multi-channel-alerting-systems.md [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+- DAG-level success or failure alerts can be created using either Airflow notifications or Astro alerts, but on Astro they are easy to configure and don't require any code changes. To implement these alerts with Airflow, you need to update your DAG code and complete additional configuration for your communication channel (for example, setting up an SMTP server for email alerts).
```
Comment (paolaperaza) [inline astro/use-cases/multi-channel-alerting-systems.md:50]: Did these need to be bold also

File: astro/use-cases/multi-channel-alerting-systems.md [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+- Absolute time alerts are recommended when you need your DAG to complete by a certain time of day. These types of alerts are available only on Astro.
```
Comment (paolaperaza) [inline astro/use-cases/multi-channel-alerting-systems.md:51]: Maybe say explicitly that they are similar to Airflow SLAs but superior

File: astro/use-cases/multi-channel-alerting-systems.md [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+
+In some cases, whether you set up an alert on Astro or in Airflow will depend on how your DAGs were configured before you moved to Astro. For example, if all of your DAGs already have email-on-failure notifications configured, it will likely be easier for you to configure an SMTP server on Astro and rely on the existing notification than implement a new Astro alert for each DAG. On the other hand, if you are developing new DAGs on Astro, it will likely be easier to implement Astro alerts which require less configuration for the communication channels.
+
+:::info
+
+Programmatically creating Astro alerts, which may be helpful if you have a large number of DAGs, is coming soon.
```
Comment (paolaperaza) [inline astro/use-cases/multi-channel-alerting-systems.md:57]:
```suggestion
Programmatically creating Astro alerts is coming soon. This will make it easy to configure alerts for a large number of DAGs without  having to create each alert manually.
```

File: astro/use-cases/multi-channel-alerting-systems.md [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+
+:::
+
+In other cases, Astro alerts are always the better option. If you need to implement SLAs for your pipelines so you immediately know when a DAG task is taking too long to run, you should use the absolute time or task duration Astro alerts. There are a couple of reasons these are preferable to Airflow SLAs:
```
Comment (paolaperaza) [inline astro/use-cases/multi-channel-alerting-systems.md:61]: Should this explanation be up top?

**Review by paolaperaza — COMMENTED** [review]
This is awesome content. No obligation to block this PR given my feedback, but my overriding thought was on the actual structure of the doc! There's so much substance and I think we can do a better job of taking the user through it.

- It wasn't clear at the top of the doc that this guide was going to help you choose between Astro Alerts and Airflow Alerts. This should be clear in the overview 
- I would honestly consider renaming to "Astro vs. Airflow Alerts"
- I would put the Airflow vs. Astro alerts explanation up higher and then end with the procedure


Hope that makes sense. It's all in here!!!!

File: astro/use-cases/multi-channel-alerting-systems.md (hunk 16 above)
Comment (kentdanas) [reply astro/use-cases/multi-channel-alerting-systems.md:51]: These aren't similar to Airflow SLAs - Airflow (right now, anyway) does not have the concept of an absolute time SLA. Airflow SLAs are more similar to our task duration alert, which we can add a bullet for.

File: astro/use-cases/multi-channel-alerting-systems.md [hunk 19]
```diff
@@ -0,0 +1,86 @@
+---
+title: 'Set up a multi-channel alerting system for your Airflow pipelines on Astro'
+sidebar_label: 'Multi-channel alerting systems'
```
[the remark this answers is not in the document]
Comment (kentdanas) [reply astro/use-cases/multi-channel-alerting-systems.md:3]: Good feedback - going to update to "Airflow vs Astro alerts". I think we made it more action-oriented because of the "use case" header, but honestly I think we should update that to "best practices" anyway. Different convo, but this would fit better into that.

File: astro/use-cases/airflow-vs-astro-alerts.md [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+To implement this use case:
+
+1. Open the DAG in your Astro project and configure a pre-built SlackNotifier for one of its tasks as a task-level argument. See [Example pre-built notifier: Slack](https://docs.astronomer.io/learn/error-notifications-in-airflow#example-pre-built-notifier-slack) for sample code.  
+2. Deploy your project to your Astro Deployment. See [Deploy code to Astro](https://docs.astronomer.io/astro/deploy-code).
+3. Add a connection to Slack in the Astro Cloud UI. See [Create Airflow connections in the Cloud UI](create-and-link-connections.md). This connection will be used by your SlackNotifier, so make sure the connection ID matches what you used in your DAG code in Step 1.
+4. In your Astro Deployment, configure an Astro **DAG failure** alert for your DAG using the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).
+5. Configure an Astro **Absolute time** alert for your DAG based on the amount of time you expect your DAG to complete in. Use the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).
+
+
```
Comment (paolaperaza) [inline astro/use-cases/airflow-vs-astro-alerts.md:90]:
```suggestion
```

**Review by paolaperaza — APPROVED** [review]
@kentdanas This is SOOOO much better! It was great before of course, but the polish and restructure here I think gives it a ton more clarity and is very easy/enjoyable to read through.

Thanks for taking the time to do that even when it was almost ready for merge <3

File: astro/use-cases/airflow-vs-astro-alerts.md [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+- **Absolute time alerts** are recommended when you need your DAG to complete by a certain time of day. These types of alerts are available only on Astro.
+- **Task duration alerts** are recommended when you need to know if your task has taken longer than a certain amount of time to complete. These types of alerts are easy to implement with Astro alerts, but are unintuitive in Airflow.
+
+In some cases, whether you set up an alert on Astro or in Airflow will depend on how your DAGs were configured before you moved to Astro. For example, if all of your DAGs already have email-on-failure notifications configured, it will likely be easier for you to configure an SMTP server on Astro and rely on the existing notification than implement a new Astro alert for each DAG. On the other hand, if you are developing new DAGs on Astro, it will likely be easier to implement Astro alerts which require less configuration for the communication channels.
+
+:::info
+
+Programmatically creating Astro alerts is coming soon. This will make it easy to configure alerts for a large number of DAGs without  having to create each alert manually.
+
+:::
```
[the remark this answers is not in the document]
Comment (kentdanas) [reply astro/use-cases/airflow-vs-astro-alerts.md:43]: I intentionally put it directly following the discussion on whether to use Airflow vs Astro for things where functionality is supported on either, because I'm arguing that for DAG failure/success alerts, the only reason to use Airflow is if you already have them set up so you don't have to manually add Astro alerts, so it was directly related to that guidance.
